libminifi/src/c2/C2Client.cpp

```diff
@@ -0,0 +1,318 @@
+      if (nullptr != rep) {
+        std::vector<std::shared_ptr<state::response::ResponseNode>> metric_vector;
+        rep->getResponseNodes(metric_vector);
+        std::lock_guard<std::mutex> guard(metrics_mutex_);
+        for (auto& metric : metric_vector) {
+          component_metrics_[metric->getName()] = metric;
+        }
+      }
+    }
+  }
+
+  loadC2ResponseConfiguration("nifi.c2.root.class.definitions");
+
+  if (!initialized_) {
+    c2_agent_ = std::unique_ptr<c2::C2Agent>(new c2::C2Agent(controller, update_sink, configuration_, filesystem_));
+    c2_agent_->start();
+    initialized_ = true;
+  }
+}
+
+void C2Client::loadC2ResponseConfiguration(const std::string &prefix) {
+  std::string class_definitions;
+  if (!configuration_->get(prefix, class_definitions)) {
+    return;
+  }
+
+  std::vector<std::string> classes = utils::StringUtils::split(class_definitions, ",");
+
+  for (const std::string& metricsClass : classes) {
+    try {
+      std::string option = prefix + "." + metricsClass;
+      std::string classOption = option + ".classes";
+      std::string nameOption = option + ".name";
+
+      std::string name;
+      if (!configuration_->get(nameOption, name)) {
+        continue;
+      }
+      std::shared_ptr<state::response::ResponseNode> new_node = std::make_shared<state::response::ObjectNode>(name);
+      if (configuration_->get(classOption, class_definitions)) {
+        std::vector<std::string> classes = utils::StringUtils::split(class_definitions, ",");
+        for (const std::string& clazz : classes) {
+          // instantiate the object
+          auto ptr = core::ClassLoader::getDefaultClassLoader().instantiate(clazz, clazz);
+          if (nullptr == ptr) {
+            bool found_metric = [&] {
+              std::lock_guard<std::mutex> guard{metrics_mutex_};
+              auto metric = component_metrics_.find(clazz);
+              if (metric != component_metrics_.end()) {
+                ptr = metric->second;
+                return true;
+              }
+              return false;
+            }();
+            if (!found_metric) {
+              logger_->log_error("No metric defined for %s", clazz);
+              continue;
+            }
+          }
+          auto node = std::dynamic_pointer_cast<state::response::ResponseNode>(ptr);
+          std::static_pointer_cast<state::response::ObjectNode>(new_node)->add_node(node);
+        }
+
+      } else {
+        std::string optionName = option + "." + name;
+        auto node = loadC2ResponseConfiguration(optionName, new_node);
+      }
+
+      std::lock_guard<std::mutex> guard{metrics_mutex_};
+      root_response_nodes_[name] = new_node;
+    } catch (...) {
+      logger_->log_error("Could not create metrics class %s", metricsClass);
+    }
+  }
+}
+
+std::shared_ptr<state::response::ResponseNode> C2Client::loadC2ResponseConfiguration(const std::string &prefix, std::shared_ptr<state::response::ResponseNode> prev_node) {
```

Comment:
This seems very similar to the other overload. Maybe it would be worth at least extracting the overlapping parts.

Reply:
thought about it too, but decided not to as I do not fully understand the logic behind these methods, possibly in an upcoming PR